### ISSUE TITLE
Large number of transient memory allocations during Eager test run

### ISSUE THREAD

**Issue by hassila** [mzaks/FlatBuffersSwift#25]
Just run an analysis of 1M Eager run to see malloc profile - can see a number of nice multiples of 1M small allocations, adding this issue as a placeholder for digging into if some could be eliminated (even though it is just transient allocations, they seems to be the root cause holding back performance).

Attaching screenshot from run.
<img width="1025" alt="malloc" src="https://cloud.githubusercontent.com/assets/8501048/14935486/749f1d96-0ed3-11e6-9864-08b7997c2238.png">

**Comment by hassila** [comment]
Retaining the capacity of the caches in the generated code made a noticeable improvement and improved encoding performance a further 25%.

12562 ms encode --> 10072 ms encode

I just manually made the following change in clearCaches throughout, but would need to be fixed in the codegen:

```
    static func clearCaches(){
        cache.removeAll(keepCapacity: true)
        inProgress.removeAll(keepCapacity: true)
        deferedBindings.removeAll(keepCapacity: true)
    }
```

<img width="883" alt="malloc2" src="https://cloud.githubusercontent.com/assets/8501048/14935522/0204a63c-0ed5-11e6-8b53-a2ac8a44d9db.png">

**Comment by mzaks** [comment]
`keepCapacity` is a good catch.
I thought that the caches will make encoding slower, I implemented them for #14 
Right now I create them for every table. I will implement an analyser in the code generator so it is only generated for tables which are actually recursive. E.g. the bench should have no caching because there are no recursive definitions.

Hm I just realised that I also should clear `objectPool` as it affects repeated reading.

**Comment by hassila** [comment]
The other issue with statics are thread safety of the caches... It'd be nice to be able to encode/decode in multiple threads in parallel.

**Comment by mzaks** [comment]
I think parallelisation of encoding is not useful, because buffer writing to buffer is the main bottleneck and can be done only in serial manner, as we need to now the current _cursor_ value. But if you have multiple object graphs you need to encode this can totally be done on multiple threads. Same for decoding. I already thought about that the key of the caches has to be tuple of offset and the objectId of reader/builder. This way it is still not thread safe but you can have parallelisation for distinguished graphs even if they use same tables. What do you think?

**Comment by hassila** [comment]
I was thinking about parallelization of independent object graphs only. The current code would not allow for threading as the type property (static) caches would not be thread safe (would need to synchronize over a dispatch queue for those). It would probably be better to move the caches to the respective builder/reader object to avoid the need for synchronization instead (and adjust the keys as needed) and be clear that a given object graph can only be serialized/deserialized form a single thread (which makes sense).

**Comment by mzaks** [comment]
Hm, I will meditate on that :)

**Comment by hassila** [comment]
(or drop the statics completely... I don't know exactly how much improvement they bring though.)

**Comment by mzaks** [comment]
They are not for performance optimisation but for being able to recreate a graph. For example in my find city project I have table with two arrays. those arrays reference the same objects. If we don't have `objectPool` we would duplicated every entry on eager decoding. As a matter of fact if you would convert this flat buffer to a JSON all entries would be double, as JSON is a tree. FlatBuffers lets us serialise real object graphs because everything is reference based. `cache`, `inProgress`, `deferedBindings` where introduced to being able to encode a cyclical graph. For example:

```
table Friend {
    name : string;
    friends : [Friend];
    father : Friend;
    mother : Friend;
}
root_type Friend;
```

Than we can do something like:

``` swift
let f1 = Friend()
f1.name = "Maxim"

let f2 = Friend()
f2.name = "Joakim"

f1.friends = [f2]
f2.friends = [f1]

f1.toByteArray
```

This would not be possible without `deferedBindings` because when we encode f1, we see that we have to encode f2 first, but when we encode f2, we see that we need to encode f1 first.  The `inProgress` set tells us that f1 is already in progress so encode f2 and put 0 as a reference for now. Than when everything is done in `performLateBindings` we can fill the 0 references with real offsets.

This feature is actually not even possible in the original flatc as far as I know, however I fill like it is a very powerful one and it reflects how amazing FlatBuffers is. :)

I guess we could put the caches in the reader and builder, I remember why I didn't do it in the first place. Because reader and builder have no idea about the types, but I guess we could use `AnyObject` and cast it in the table classes.

**Comment by hassila** [comment]
Could just define a protocol and type the caches on that instead of AnyObject?

**Comment by mzaks** [comment]
Yeah we could have a ghost protocol `Table`

**Comment by mzaks** [comment]
Pushed the new code generator which does not add all the recursion overhead if table definitions are not recursive. Also removed static fields, they are on the builder and reader now. I guess this issue can be closed than.

**Comment by hassila** [comment]
It still does try to access the pools from the non-recursive benchmark and there are code generated like:

```
public extension Foo {
    private static func create(reader : FlatBufferReader, objectOffset : Offset?) -> Foo? {
        guard let objectOffset = objectOffset else {
            return nil
        }
-->     if let o = reader.objectPool[objectOffset]{
            return o as? Foo
        }
        let _result = Foo()
-->     reader.objectPool[objectOffset] = _result
        _result.id = reader.get(objectOffset, propertyIndex: 0, defaultValue: 0)
        _result.count = reader.get(objectOffset, propertyIndex: 1, defaultValue: 0)
        _result.prefix = reader.get(objectOffset, propertyIndex: 2, defaultValue: 0)
        _result.length = reader.get(objectOffset, propertyIndex: 3, defaultValue: 0)
        return _result
    }
}

```

Also, as I link with the library, that code can't actually access the properties as they are in a different module, created https://github.com/mzaks/FlatBuffersSwift/pull/26 to fix that.

**Comment by hassila** [comment]
(on a side note just FYI - running with the code generated inline rather than linking with the framework lets the optimizer do its magic - encoding becomes almost 2x faster...)

**Comment by hassila** [comment]
Also:

```
public extension Foo {
    private func addToByteArray(builder : FlatBufferBuilder) -> Offset {
-->     if let myOffset = builder.cache[ObjectIdentifier(self)] {
            return myOffset
        }
        try! builder.openObject(4)
        try! builder.addPropertyToOpenObject(3, value : length, defaultValue : 0)
        try! builder.addPropertyToOpenObject(2, value : prefix, defaultValue : 0)
        try! builder.addPropertyToOpenObject(1, value : count, defaultValue : 0)
        try! builder.addPropertyToOpenObject(0, value : id, defaultValue : 0)
        let myOffset =  try! builder.closeObject()
-->     builder.cache[ObjectIdentifier(self)] = myOffset
        return myOffset
    }
}

```

I tried manually removing them just to see what the cache overhead was, quite significant:

```
Eager run
=================================
8605 ms encode
6966 ms decode
1037 ms use
1504 ms dealloc
9507 ms decode+use+dealloc
=================================

—>

=================================
6239 ms encode
4282 ms decode
994 ms use
1449 ms dealloc
6725 ms decode+use+dealloc
=================================
```

**Comment by hassila** [comment]
Even so, the number of mallocs actually significantly increased: <img width="996" alt="mallocs" src="https://cloud.githubusercontent.com/assets/8501048/14940485/b3527bf6-0f77-11e6-8436-6ee9210b25e5.png">

**Comment by hassila** [comment]
Roughly 64M instead of 38M allocations. Did a test run removing the "stringCache", which removed 4M allocations and improved eager decoding performance ~30% (4447->3038) - did you see better results by adding it initially?

**Comment by mzaks** [comment]
`stringCache` wa always slowing down the performance tests, I even removed it for some time. It slows it down because with randomly generated string you always have cache misses. With real world scenario however it helped and made the size of the resulting binary much smaller. I think it was about 25% less. Repeating strings is a very common thing in real world scenarios. This cache makes sure that you encode the string once and everyone reference only this one string. 
`builder.cache` serves the same purpose but for tables. If you have same table referred from different places, this table will be encoded only once. This is also why I kept it, it is very important for directed cyclic graph, but also for directed acyclic graph, as you can have one node referenced from multiple other nodes. And again in performance test we end up with only cache misses.

So I gutes it is time for the **ConfigObject**. It will have all the caches turned on per default, but user can turn it off when s/he knows that they are doing more harm than good in a special use case. It will also contain the initial capacity. However I am not sure if I will be able to work on it today. Let's see.

**Comment by hassila** [comment]
Ok, but in this benchmark the strings used are not randomly generated, there are only two constant strings, so then likely something is not working as expected? 

I guess I am primarily focusing on simple object graphs without cross references, so tuning it in a config would be good enough for practical purposes.

One concern with the config object though is that the reader/builder instance variables still causes 10M allocations just by their existence (tried removing them as well, gave a nice speedup of 10%).

Thinking about it, wouldn't it make sense to be able to reset builder/readers and reuse them instead? Then we could live with a small amount of instances of those and it would essentially be a non-issue with those variables.

**Comment by mzaks** [comment]
And the configs are in, give it a try. I set `uniqueStrings` to `false`. You can try to turn them on. I also saw that the builder was lacking the `stringCache` which was a dumb mistake.

**Comment by hassila** [comment]
Ok, will give it a spin - what do you think about trying to reuse builers/readers ?

**Comment by hassila** [comment]
Using the string cache shifted time from decode to build as expected, but overall nice option I think:

```
String cache OFF:
Lazy run
=================================
7295 ms encode
386 ms decode
5246 ms use
1013 ms dealloc
6645 ms decode+use+dealloc
=================================

Eager run
=================================
7515 ms encode
3667 ms decode
1176 ms use
1675 ms dealloc
6518 ms decode+use+dealloc
=================================

String cache turned ON:
Lazy run
=================================
8194 ms encode
364 ms decode
4701 ms use
1108 ms dealloc
6173 ms decode+use+dealloc
=================================

Eager run
=================================
8121 ms encode
3209 ms decode
1047 ms use
1335 ms dealloc
5591 ms decode+use+dealloc
=================================

```

**Comment by hassila** [comment]
FYI, I just did hack in a reuse of the reader (manually added):

```
    public static func fromReader(r : FlatBufferReader) -> FooBarContainer {
        let reader = r
        let objectOffset = reader.rootObjectOffset
        return create(reader, objectOffset : objectOffset)!
    }

```

and just manually reset the buffer pointer between runs as a prototype (should have a nice reset method of course), which gave a significant boost really:

```
Eager run
=================================
7576 ms encode
3299 ms decode
1029 ms use
1468 ms dealloc
5796 ms decode+use+dealloc
=================================

—>

Eager run
=================================
8040 ms encode
1207 ms decode
1089 ms use
1200 ms dealloc
3496 ms decode+use+dealloc
=================================

```

Decode time was cut to 1/3 or so :)

**Comment by hassila** [comment]
Tried the same for builders, reduced total allocations down to 40M (down from 60+) and gave a similar boost on encode performance:

```
Eager run
=================================
5377 ms encode
1112 ms decode
998 ms use
1095 ms dealloc
3205 ms decode+use+dealloc
=================================

```

So seems worthwhile to consider to allow reusing builders/readers. 40M allocations is also much more reasonable vs the workload.

**Comment by mzaks** [comment]
Interesting!
I thought that builder and reader reuse is not significant in real life scenario, but this is the FrontEnd developer in me talking :)
If a BackEnd has to consume and produce slightly different buffers this is indeed useful.
#17 goes kind of in the same direction, but it is much more limited because there it is only possible to exchange scalar values, and there is also a complication with default values btw.
#13 should also speed up encoding, but on cost of nice API.

I created a new issue :) #27

**Comment by hassila** [comment]
In my experience (mostly doing backend stuff) every allocation counts - really why I added this issue to be able to ensure that the foundation is doing a reasonable job there so it can be used on the backend side (as a bonus it would work nicer on mobile devices also :) ). 

Looking at the test run, there is an object graph of 10 objects with 2 strings and one vector (13 objects total), so ~13+13M=26M allocations as a floor for the test run would not be unreasonable - if reusing the reader/builder (#27) we have a surplus of ~14M unaccounted for which still would be nice to sort out if possible. I would hope that building/reading performance could come close to the reference C++ implementation and while the gap is closing, there is bit to go.

#17 was created with the idea that the backend store could be kept around for Lazy access, perhaps that would clash with the idea of reuse somewhat.

Not sure how #13 would look like, but I do like the current API (modulo #27 ;-)

I think the major question is - do one have a transient object graph, or a persisted one that is updated and sent repeatedly - I can see uses for both for different scenarios that are useful.

**Comment by hassila** [comment]
See https://github.com/mzaks/FlatBuffersSwift/issues/28 for one more cheap win.

**Comment by hassila** [comment]
And one more in https://github.com/mzaks/FlatBuffersSwift/pull/29

**Comment by hassila** [comment]
It seems one major issue is the withUnsafePointer closure in public func put<T : Scalar>(value : T), not sure what could be done there though.

**Comment by hassila** [comment]
That seems to have been a red herring, it seems to be the vtable vectors in the builder that is the major culprit. Lets close this case and I'll open separate ones when nailing down concrete suggestions.
